Laser envelope solver (finally!)

MaxThevenet opened the pull request:
This PR proposes an implementation of a laser envelope solver. The plasma response to an analytic laser pulse was already implemented. Now, the propagation of the laser pulse in a plasma is also included. The model is based on [Benedetti's 2017/2018 article](https://iopscience.iop.org/article/10.1088/1361-6587/aa8977/meta?casa_token=UwsLmt5ouXwAAAAA:SeLUPCyDq4xPyJHw-gii5ow4AGaKbwzIn8MkIUckiNyhUk7cYkYJfFhEQZNefn4Kmo8ndQX87w).

## Features
- The complex envelope of current and previous time steps is stored as a 3D array (currently with same bounds and resolution).
- The 3D array can be stored on host or on device memory, as chosen by a runtime parameter. The copies from/to the 3D array is implemented.
- Runs on Nvidia GPUs (and CPUs).
- Runs in parallel.
- The envelope is dumped in openPMD files.
- A CI test checks the evolution of a laser pulse in vacuum (and compares with theory).

## Structure

The 3D array is `Laser::m_F`; 2D slices (`FArrayBox`es) are stored in `Laser::m_slices`, labelled `n**j**` where n stands for the time step, j for the longitudinal slice. n00 is time step n, nm1 is n-1 and np1 is n+1. A similar notation is used for slice j. In general, fields are stored with 2 `Real` components, one for the real part, the other for the imaginary part, except inside the FFT solver (where fields are directly stored as `Complex` arrays).

A new Poisson solver is implemented to solve the C2C Poisson equation with periodic BC. Although this causes duplication with the existing solver, it reduces complexity (the capability to abstract the type of the source array of the forward FFT to be either `Real`, as needed everywhere, or `Complex`, as needed for the laser pulse, would take significant templating and obfuscate the code). This could be reconsidered. The solver is directly implemented in `Laser.cpp` in `Laser::AdvanceSliceFFT`, with the required abstraction for portability in `Laser.H`.

Two solvers are implemented, a MG solver and a FFT solver, to advance a laser slice by 1 time step. This solver operation computes slice `j` at step `n+1` using previous slices and previous time steps: $s_{j}^{n+1} = f(s_{j+1,j+2}^{n+1}, s_{j,j+1,j+2}^{n}, s_{j,j+1,j+2}^{n+1})$. This is integrated within the loop over slices. The management of these slices is largely done in `Laser::Copy`.

For parallel runs, the whole 3D array on the current box has to be communicated. This is done similar to beam communication. 

A new quantity `chi` has to be deposited (in `PlasmaCurrentDepositionInner.H`) for the plasma response. This is essentially the plasma density divided by the Lorentz factor. It is used in the laser solver.

## Performance

The CI test with resolution `1024 1024 500` for 7 time steps gives the following time (in vacuum, so the changes would not be that dramatic with a plasma):
```
$ grep "total time" output*txt
output.old_ranks.1_host.0.txt:TinyProfiler total time across processes [min...avg...max]: 17.49 ... 17.49 ... 17.49
output.old_ranks.1_host.1.txt:TinyProfiler total time across processes [min...avg...max]: 35.19 ... 35.19 ... 35.19
output.old_ranks.4_host.0.txt:TinyProfiler total time across processes [min...avg...max]: 20.34 ... 21.88 ... 23.19
output.old_ranks.4_host.1.txt:TinyProfiler total time across processes [min...avg...max]: 26.69 ... 28.66 ... 30.36
```
and memory usage
```
$ grep "Free  GPU global memory" output*txt
output.old_ranks.1_host.0.txt:Free  GPU global memory (MB) spread across MPI: [23391 ... 23391]
output.old_ranks.1_host.1.txt:Free  GPU global memory (MB) spread across MPI: [39361 ... 39361]
output.old_ranks.4_host.0.txt:Free  GPU global memory (MB) spread across MPI: [35265 ... 35265]
output.old_ranks.4_host.1.txt:Free  GPU global memory (MB) spread across MPI: [39225 ... 39353]
```
when changing the number of ranks and whether the laser envelope is on host or device. As expected, storing the 3D laser envelope on host makes the code slower, but uses less memory.

## Remains to be done

See #804.

MaxThevenet committed:
store 3D laser on host and do H2D & D2H copies

MaxThevenet committed:
Serial run with dummy laser advance works

MaxThevenet committed:
start parallelization, still buggy

MaxThevenet committed:
fix parallelization

MaxThevenet committed:
re-enable tiling

MaxThevenet committed:
first attempt of laser IO

MaxThevenet committed:
fix and cleaning

MaxThevenet committed:
minor

MaxThevenet committed:
Merge pull request #2 from MaxThevenet/laser2

first attempt of laser IO

MaxThevenet committed:
add a few synchronize for GPU

MaxThevenet committed:
clean up useless comments and print statements

MaxThevenet committed:
Option to allocate 3D array on host or device

MaxThevenet committed:
minor comment update

MaxThevenet committed:
Update arrays and data for realistic update: need MUCH more slices from several time steps etc.

MaxThevenet committed:
fix merge conflicts

MaxThevenet committed:
fix merge conflicts

MaxThevenet committed:
properly handly imaginary and real part of laser envelope

MaxThevenet committed:
compute all arrays needed in the MG solve for the laser advance

MaxThevenet committed:
call solber

MaxThevenet committed:
debugging: only with k0=0 and dz=inf does the MG solver converge

MaxThevenet committed:
temporary test functions for debugging

MaxThevenet committed:
tmp chance

MaxThevenet committed:
Merge branch 'development' into laser

MaxThevenet committed:
new MG solver allows to switch back to actual implementation (vacuum so far). But amplitude evolves wrongly.

MaxThevenet committed:
add new version with on-axis phase evaluation only

MaxThevenet committed:
works serial, on cpu, in vacuum

MaxThevenet committed:
start cuda implementation

MaxThevenet committed:
Fix GPU implementation

MaxThevenet committed:
conditional include for cufft stuff

MaxThevenet committed:
add guard cells to laser and fix plasma response

MaxThevenet committed:
get on-axis complex phase in a GPU-friendly way

MaxThevenet committed:
cleaning, commenting, and fix reduction operation to get on-axis phase

MaxThevenet committed:
capability to focus laser at arbitrary position

MaxThevenet committed:
add automated test for laser propagation in vacuum

MaxThevenet committed:
fix CI: laser.lambda0 is now a mandatory parameter

MaxThevenet committed:
cosmetic changes

MaxThevenet committed:
do not access laser array if laser not used

MaxThevenet committed:
reset benchmarks

MaxThevenet committed:
fix checkout

MaxThevenet committed:
remove tabs

MaxThevenet committed:
add a bit of doc

MaxThevenet committed:
update benchmark

MaxThevenet committed:
bit of doc

MaxThevenet committed:
doc

MaxThevenet committed:
merge dev

MaxThevenet committed:
additional fix to use Array3

MaxThevenet committed:
switch from Array4 to Array3

MaxThevenet committed:
remove unused array and clean up AMD compilation

MaxThevenet committed:
first attempt to incorporate plasma response to laser evolution

MaxThevenet committed:
fix plasma response to laser for FFT solver

MaxThevenet committed:
reset all laser slices. Still instability growing for long propagation

MaxThevenet committed:
laser slices must be reset when starting a new time step

MaxThevenet committed:
fix FFT slice advance and use Complex numbers all along

MaxThevenet committed:
Fix laser MG solver. Imag RHS still single number

MaxThevenet committed:
fix AMD compilation

MaxThevenet committed:
Doxygen fix

MaxThevenet committed:
do not allocate chi if laser not used

MaxThevenet committed:
add chi to checksum benchmarks when missing

MaxThevenet committed:
clean 1/2 step to remove spurious numerical effect during propagation

MaxThevenet committed:
add comment

MaxThevenet committed:
ups

MaxThevenet committed:
cleaner handling of 3d-slice copies and laser slices reset. 2 ranks look ballpark correct, still a bug causing noise around the box boundaries

MaxThevenet commented:
TO do now:
- ~~Fix parallelization~~ should be fine now. Small differences still (1.e-9), maybe communicating guard cells would fix it

To do later:
 - Fix Array3/Array4 thing
 - Implement AMD
 - Check on GPU
 - Properly handle Mult and Chi

MaxThevenet committed:
fix parallelization: slice np1j00 was wrongfully overwritte during init

MaxThevenet committed:
fix merge conflicts

MaxThevenet committed:
fix AMD compilation

MaxThevenet committed:
remove unused variable

MaxThevenet committed:
laser CI on 2 ranks, and with checksum benchmark

MaxThevenet committed:
explicit flag to turn on laser

MaxThevenet committed:
fix checksum comparison

MaxThevenet committed:
cleaning and update doc

MaxThevenet committed:
make sure only 1 plasma species. >1 not yet supported

MaxThevenet committed:
typo

MaxThevenet committed:
doc

MaxThevenet committed:
clean up error message and doc for solver type

MaxThevenet committed:
remove unused parameter in Laser::Copy

SeverinDiederichs commented on docs/source/run/parameters.rst:
```diff
@@ -497,11 +497,11 @@ Option: ``from_file``
+The laser pulse length :math:`L0 = \tau / c_0` can be specified via the pulse duration ``laser.tau``.
+
+* ``laser.use_laser`` (`0` or `1`) optional (default `0`)
+    Weather to activate the laser envelope solver.
```
```suggestion
    Whether to activate the laser envelope solver.
```

Could we make the default to 1 if a laser is specified in the input script, i.e., a0 > 0?

SeverinDiederichs replied:
Should we explain that we do not include direct laser acceleration?

MaxThevenet replied:
I like keeping this parameter explicit, so it is easy to de-activate.
And for an envelope solver, I think it is expected that there is no DLA.

SeverinDiederichs commented on examples/laser/inputs_SI:
```diff
@@ -0,0 +1,50 @@
+hipace.file_prefix = new
+hipace.do_tiling = 0
+
+geometry.prob_lo     = -6.*kp_inv   -6.*kp_inv   -8.*kp_inv
```
Could we move the definition of kp_inv above this?

SeverinDiederichs commented on examples/laser/inputs_SI:
```diff
@@ -0,0 +1,50 @@
+
+amr.n_cell = 128 128 50 # 128 128 100
+
+# plasmas.sort_bin_size = 8
```
```suggestion
```

MaxThevenet replied:
yes

SeverinDiederichs commented on examples/laser/inputs_SI:
```diff
@@ -0,0 +1,50 @@
+plasma.density(x,y,z) = ne
+plasma.ppc = 1 1
+plasma.u_mean = 0.0 0.0 0.
+plasma.element = electron
```
```suggestion
```

(unused)

SeverinDiederichs commented on examples/laser/inputs_SI:
```diff
@@ -0,0 +1,50 @@
+my_constants.kp_inv = 10.e-6
+my_constants.kp = 1. / kp_inv
+my_constants.wp = clight * kp
+my_constants.ne = wp^2 * m_e * epsilon0 / q_e^2
```
```suggestion
```

(unused)

SeverinDiederichs commented on docs/source/run/parameters.rst:
```diff
@@ -519,8 +519,16 @@ If no ``laser.a0`` is provided, no laser will be initialized.
+* ``laser.focal_distance`` (`float`)
+    Distance at which the laser pulse if focused (in the z direction, counted from laser initial position).
+
+* ``laser.solver_type`` (`string`) optional (default `fft`)
```
What is the reason for making FFT default over MG? Was it faster?

MaxThevenet replied:
It should be faster and more accurate, this is currently just a complex Poisson equation.

SeverinDiederichs commented on src/Hipace.cpp:
```diff
@@ -426,6 +431,16 @@ Hipace::Evolve ()
+            const amrex::Box& bx = boxArray(lev)[it];
+
+            if (m_laser.m_use_laser) {
+                AMREX_ALWAYS_ASSERT(!m_adaptive_time_step.m_do_adaptive_time_step);
```
Could you add the adaptive time step for the laser to the to do list in the PR please?

MaxThevenet replied:
ok, I'll put it down the list, this would take a bit of thinking.

SeverinDiederichs commented on src/Hipace.cpp:
```diff
@@ -1413,6 +1482,30 @@ Hipace::Notify (const int step, const int it,
+        if (only_ghost) return;
+        if (!m_laser.m_use_laser) return;
+        const amrex::FArrayBox& laser_fab = m_laser.getFAB();
+        // Note that there is only one local Box in slice multifab's boxarray.
+        // const int box_index = laser_fab.IndexArray();
```
```suggestion
        // Note that there is only one local Box in slice multifab's boxarray.
        // const int box_index = laser_fab.IndexArray();
```

Are these comments outdated?

MaxThevenet replied:
probably

SeverinDiederichs commented on docs/source/run/parameters.rst:
```diff
@@ -519,8 +519,16 @@ If no ``laser.a0`` is provided, no laser will be initialized.
+    Type of solver for the laser envelope solver, either ``fft`` or ``multigrid``.
+    Currently, the approximation that the phase is evaluated on-axis only is made with both solvers.
+    With the multigrid solver, we could drop this assumption.
 
```
We should add the `3d_on_host` parameter to this list.

MaxThevenet replied:
indeed

SeverinDiederichs commented on src/fields/Fields.cpp:
```diff
@@ -482,6 +494,18 @@ Fields::Copy (const int lev, const int i_slice, const amrex::Geometry& diag_geom
                 const int m = n[diag_comps];
                 diag_array(i,j,k,n) += rel_z_data[k-k_min] * slice_array(x,y,m);
             });
+        amrex::Gpu::Device::synchronize();
```
Are the synchronizes needed here? I thought that after a `ParallelFor` it is always synchronized.

MaxThevenet replied:
Probably superfluous indeed

SeverinDiederichs commented on src/laser/Laser.H:
```diff
@@ -1,44 +1,169 @@
     const amrex::MultiFab& getSlices (int islice) const {return m_slices[islice]; }
 
-    /** \brief Allocate laser multifab */
+    /** Retun the 3D FArrayBox containing the laser envelope, const version */
```
```suggestion
    /** Return the 3D FArrayBox containing the laser envelope, const version */
```

SeverinDiederichs commented on src/laser/Laser.H:
```diff
@@ -1,44 +1,169 @@
+    /** Retun the 3D FArrayBox containing the laser envelope, const version */
+    const amrex::FArrayBox& getFAB () const {return m_F;}
+
+    /** Retun the 3D FArrayBox containing the laser envelope, non-const version */
```
```suggestion
    /** Return the 3D FArrayBox containing the laser envelope, non-const version */
```

SeverinDiederichs commented on src/laser/Laser.H:
```diff
@@ -1,44 +1,169 @@
+    void AdvanceSlice (const Fields& fields, const amrex::Geometry& geom, amrex::Real dt, int step);
+
+    /** Advance a laser slice by 1 time step using a multigrid solver.
+     * The complex phase of the envelope is evaluated everywhere.
```
This statement contradicts the `parameters.rst`, where it says:
```
Currently, the approximation that the phase is evaluated on-axis only is made with both solvers.
With the multigrid solver, we could drop this assumption.
```

SeverinDiederichs commented on src/laser/Laser.H:
```diff
@@ -1,44 +1,169 @@
 /** \brief describes which slice with respect to the currently calculated is used */
 struct WhichLaserSlice {
-    enum slice { This=0, PrevZeta1, PrevZeta2, PrevTime1, PrevTime2, N };
+    // n stands for the time step, j for the longitudinal slice.
+    // n00 is time step n, nm1 is n-1 and np1 is n+1. Similar notation for slice j.
+    enum slice { nm1j00, nm1jp1, nm1jp2, n00j00, n00jp1, n00jp2, np1j00, np1jp1, np1jp2, N };
```
Is there a specific reason for n and j as indices? Could we use t for time and s for slice, so `tp1sm1`?

MaxThevenet replied:
This is standard notation for envelope solver

SeverinDiederichs commented on src/laser/Laser.cpp:
```diff
@@ -35,70 +55,671 @@ Laser::InitData (const amrex::BoxArray& slice_ba,
+    for (int isl = bx.bigEnd(Direction::z); isl >= bx.smallEnd(Direction::z); --isl){
+        // Compute initial field on the current (device) slice n00j00 for current
+        InitLaserSlice(gm, isl);
+        // Copy: (device) slice to (host) 3D array. A = np1j00
```
What does `A = np1j00` mean?

MaxThevenet replied:
This was clarified

SeverinDiederichs commented on src/laser/Laser.H:
```diff
@@ -47,14 +172,45 @@ private:
+    amrex::FArrayBox m_F;
+    /** Array of N slices required to compute current slice */
     std::array<amrex::MultiFab, m_nslices> m_slices;
+    amrex::Real m_MG_tolerance_rel = 1.e-4;
```
This should be an input parameter just like in the explicit solver. In single precision and high resolution, 1.e-4 can be too restrictive, so one has to go up to 1e-3 or something like that

MaxThevenet replied:
indeed

SeverinDiederichs commented on src/laser/Laser.cpp:
```diff
@@ -35,70 +55,671 @@ Laser::InitData (const amrex::BoxArray& slice_ba,
+    amrex::Geometry slice_geom = amrex::Geometry(
+        tmp_dom, tmp_probdom, geom.Coord(), geom.isPeriodic());
+
+    slice_geom.setPeriodicity({0,0,0});
```
does that mean that the MG solver does not have periodic BC, while the FFT solver has?

MaxThevenet replied:
Yes I assume so. The FFT solver could be improved to have better BC (Dirichlet or open BC).

SeverinDiederichs commented on src/laser/Laser.cpp:
```diff
@@ -35,70 +55,671 @@ Laser::InitData (const amrex::BoxArray& slice_ba,
+                        + ( -3._rt/(c*dt*dz) + 2._rt*I*djn/(c*dt) + 2._rt/(c*c*dt*dt) + I*2._rt*k0/(c*dt) ) * anm1j00;
+                }
+                rhs_arr(i,j,0) = rhs;
+            });
```
Up to here it is the same as in the MG solver, should we outsource this in some preparation function?

MaxThevenet replied:
There are actually some differences. We could abstract them away, but a bit more tests would probably be needed to see what is most useful on the long run. Therefore, I would like to keep it that way.

SeverinDiederichs commented on src/laser/Laser.cpp:
```diff
@@ -35,70 +55,671 @@ Laser::InitData (const amrex::BoxArray& slice_ba,
+        // Transform rhs to Fourier space
+#ifdef AMREX_USE_CUDA
+        cudaStream_t stream = amrex::Gpu::Device::cudaStream();
+        cufftSetStream ( m_plan_fwd, stream);
```
```suggestion
        cufftSetStream( m_plan_fwd, stream);
```

SeverinDiederichs commented on src/laser/Laser.cpp:
```diff
@@ -35,70 +55,671 @@ Laser::InitData (const amrex::BoxArray& slice_ba,
+    // Basic laser parameters and constants
+    Complex I(0,1);
+    const amrex::Real c = get_phys_const().c;
     const int dcomp = 0; // NOTE, this may change when we use slices with Comps
```
```suggestion
    const int dcomp = 0;
```

MaxThevenet replied:
Somehow, this comment is still accurate. But indeed, it would apply to many places, so let's just remove it.

SeverinDiederichs commented on src/particles/pusher/FieldGather.H:
```diff
@@ -245,19 +245,34 @@ void doLaserGatherShapeN (const amrex::ParticleReal xp,
     for (int iz=0; iz<=depos_order_z; iz++){
```
the TODO comment can be removed

SeverinDiederichs commented on tests/laser_blowout_wake_explicit.1Rank.sh:
```diff
@@ -29,7 +29,9 @@ mpiexec -n 1 $HIPACE_EXECUTABLE $HIPACE_EXAMPLE_DIR/inputs_normalized \
         beams.names = no_beam \
         geometry.prob_lo     = -20.   -20.   -8  \
         geometry.prob_hi     =  20.    20.    6  \
+        laser.use_laser = 1 \
```
Let's remove this once the default is set automatically

MaxThevenet replied:
I think it's a bit cleaner this way.

SeverinDiederichs commented on tests/laser_blowout_wake_explicit.SI.1Rank.sh:
```diff
@@ -28,7 +28,9 @@ mpiexec -n 1 $HIPACE_EXECUTABLE $HIPACE_EXAMPLE_DIR/inputs_SI \
         beams.names = no_beam \
         geometry.prob_lo     = -20.*kp_inv   -20.*kp_inv   -8.*kp_inv  \
         geometry.prob_hi     =  20.*kp_inv    20.*kp_inv    6.*kp_inv  \
+        laser.use_laser = 1 \
```
same as above

SeverinDiederichs commented on src/utils/AdaptiveTimeStep.H:
```diff
@@ -20,12 +20,12 @@ private:
     amrex::Real m_nt_per_betatron = 40.;
 
 public:
+    /** Whether to use an adaptive time step */
+    bool m_do_adaptive_time_step = false;
```
I think this causes a warning, see offline discussion

SeverinDiederichs reviewed (commented):
Great! See a couple of comments below.

Further problems discussed offline.

A few more items to be added to the to-do list:
1. Initialization of a laser profile via the parser (non-Gaussian),
2.  the possibility to propagate the laser backwards,
3.  the possibility to load and restart the laser.

MaxThevenet committed:
fix merge conflict

MaxThevenet committed:
fix SP on CUDA with FFT Laser solver

MaxThevenet committed:
implement changes from review

MaxThevenet committed:
no wait and notify if running on only 1 rank

AlexanderSinn commented on src/laser/Laser.H:
```diff
@@ -1,44 +1,171 @@
+    using cufftComplex = cuComplex;
+#  else
+    const auto VendorExecute = cufftExecZ2Z;
+    const auto cuufft_type = CUFFT_Z2Z;
```
```suggestion
    const auto cufft_type = CUFFT_Z2Z;
```

AlexanderSinn commented on src/fields/Fields.cpp:
```diff
@@ -482,6 +494,17 @@ Fields::Copy (const int lev, const int i_slice, const amrex::Geometry& diag_geom
                 diag_array(i,j,k,n) += rel_z_data[k-k_min] * slice_array(x,y,m);
             });
+
+        if (!laser.m_use_laser) return;
```
```suggestion
        if (!laser.m_use_laser) continue;
```

AlexanderSinn commented on src/particles/pusher/FieldGather.H:
```diff
@@ -241,23 +241,37 @@ void doLaserGatherShapeN (const amrex::ParticleReal xp,
-                   -a_arr(j_cell+ix  , k_cell+iy-1, amrex::lbound(a_arr).z)*
-                    a_arr(j_cell+ix  , k_cell+iy-1, amrex::lbound(a_arr).z));
+                const amrex::Real x00y00 =
+                    a_arr(j_cell+ix  , k_cell+iy  , amrex::lbound(a_arr).z,0)*
+                    a_arr(j_cell+ix  , k_cell+iy  , amrex::lbound(a_arr).z,0)+
+                    a_arr(j_cell+ix  , k_cell+iy  , amrex::lbound(a_arr).z,1)*
+                    a_arr(j_cell+ix  , k_cell+iy  , amrex::lbound(a_arr).z,1);
```
I think a `abs(real, imag)` function would be good here.

MaxThevenet replied:
Thanks, that's included now

MaxThevenet committed:
some cleaning

MaxThevenet commented:
Alright, thanks for all the comments! I believe all are either solved or listed in the todo above.

SeverinDiederichs commented:
Could you please comment on the status of point 1 on the to do list? Was this resolved at least for serial runs now? Or is this still present?

Otherwise, I think we can merge soon and move the to do list to an issue.

MaxThevenet commented:
Point 1 above is still relevant. This is not a surprise: the Notify/Wait code full is written for `3d_on_host = 0`, and makes no sense for `3d_on_host = 1`. I'll try to dedicate some time today to fix it. If not, we can merge and take care of it in a subsequent PR.
For serial runs, it is fixed indeed with an exit condition to the wait/notify functions.

MaxThevenet committed:
more doc

MaxThevenet commented:
Option `3d_on_host` could be accelerated, but it behaves as expected (it does reduce a lot the memory footprint). I added some info on that in the PR description. Therefore, I think this PR is good to go. I split point 1 in the todo list into 2 more detailed points. I also updated the doc to mention that the MG solver is currently less stable (offline chat with @SeverinDiederichs).

SeverinDiederichs reviewed (approved):
🎉  Awesome!
Let's merge this now and move the to do list to an issue and work on it on separate PRs 🚀
